front/lib/api/assistant/configuration.ts

```diff
@@ -204,47 +226,160 @@ export async function getAgentConfiguration(
 }
 
 /**
- * Get the list agent configuration for the workspace, optionally whose names
+ * Get agent configurations for the workspace, optionally whose names
  * match a prefix
+ * @param agentsGetView the kind of list of agents we want to get, see AgentsGetViewType
  */
 export async function getAgentConfigurations(
   auth: Authenticator,
+  agentsGetView: AgentsGetViewType,
   agentPrefix?: string
 ): Promise<AgentConfigurationType[] | []> {
   const owner = auth.workspace();
   if (!owner) {
     throw new Error("Unexpected `auth` without `workspace`.");
   }
+  if (!auth.isUser()) {
+    throw new Error("Unexpected `auth` from outside workspace.");
+  }
 
-  const rawAgents = await AgentConfiguration.findAll({
-    where: {
-      workspaceId: owner.id,
-      status: "active",
-      ...(agentPrefix
-        ? {
-            name: {
-              [Op.iLike]: `${agentPrefix}%`,
-            },
-          }
-        : {}),
-    },
-    order: [["name", "ASC"]],
-  });
-
-  const agents = (
-    await Promise.all(
-      rawAgents.map(async (a) => {
-        return await getAgentConfiguration(auth, a.sId);
-      })
-    )
-  ).filter((a) => a !== null) as AgentConfigurationType[];
+  const where = {
+    workspaceId: owner.id,
+    status: "active",
+    ...(agentPrefix ? { name: { [Op.iLike]: `${agentPrefix}%` } } : {}),
+  };
 
   const globalAgents = (await getGlobalAgents(auth)).filter(
     (a) =>
       !agentPrefix || a.name.toLowerCase().startsWith(agentPrefix.toLowerCase())
   );
 
-  return [...globalAgents, ...agents];
+  const getLocalAgentsFromIds = async (agentIds: string[]) => {
+    return (
+      await Promise.all(agentIds.map((a) => getAgentConfiguration(auth, a)))
+    ).filter((a) => a !== null) as AgentConfigurationType[];
+  };
+
+  if (agentsGetView === "superuser") {
+    if (!auth.isDustSuperUser()) {
+      throw new Error("superuser view is for dust superusers only.");
+    }
+    const agentIds = (
+      await AgentConfiguration.findAll({
+        where,
+        attributes: ["sId"],
+      })
+    ).map((a) => a.sId);
+    return [...(await getLocalAgentsFromIds(agentIds)), ...globalAgents];
+  }
+
+  if (agentsGetView === "all") {
+    const agentIds = (
+      await AgentConfiguration.findAll({
+        where: {
+          ...where,
+          scope: { [Op.in]: ["published", "workspace"] },
+        },
+        attributes: ["sId"],
+      })
+    ).map((a) => a.sId);
+    return [...(await getLocalAgentsFromIds(agentIds)), ...globalAgents];
+  }
+
+  if (agentsGetView === "list") {
+    const user = auth.user();
+    if (!user) {
+      throw new Error("List view is specific to a user.");
+    }
+    const agentIds = (
+      await AgentConfiguration.findAll({
+        where: {
+          ...where,
+          [Op.or]: [
+            { scope: { [Op.in]: ["published", "workspace"] } },
```

Comment:
scope is not indexed on AgentConfiguration we should index it no?

Reply:
[score, userId] to be precise?

Reply:
I had left the following com on the first review
```
There is no index on scope, low-cardinality indices are apparently not efficient and from what I gather often ignored by the query planner

I don't think it is a problem with this kind of query though because the index on workspaceId + active will sufficientlly narrow the search, and we only return sIds + scopes with this query
```
Indexing on low-cardinality columns is discouraged afaik
=> do you still want to add an index?

Reply:
OK